### ISSUE TITLE
Compensate "missed" triggering by UZSU on smarthomeNG startup

### ISSUE THREAD

**Issue by onkelandy** [smarthomeNG/smarthome#245]
Problem: When smarthomeNG is down at a specific time when the uzsu plugin would switch an item on/off, the item doesn't have the correct value after a restart of smarthomeNG.

Would be perfect to have an evaluation of all uzsu items at startup that looks for the last set value. If smarthomeNG wasn't running at that time, the uzsu should get triggered on startup.

Specific example:
uzsu turns item on at 18.00
smarthomeNG is down for maintenance between 17.50 and 18.15. So after restart item is still off.
With the new feature smarthomeNG realizes that it missed the "turn on" at 18.00 and triggers that action after startup.

**Comment by bmxp** [comment]
This was already discussed in [knxuf](https://knx-user-forum.de/forum/supportforen/smarthome-py/1098033-uzsu-schaltzeiten-w%C3%A4hrend-reboot-nachholen) back in May:
Essentially you will have problems due to the non-linear system nature. Image you have more than one entry in UZSU within the missed time frame. So the only solution might be to take the time now and find out which point in time the last action should have take place and replay that then.
We would need to extend UZSU plugin to do calculation backwards in time. Feel free to try to accomplish that task...

**Comment by msinn** [comment]
There maybe multiple events in the time where SmartHomeNG is down. Maybe even multiple Events for the same Item. 

To be on the safer side one would have to "replay" all events that happend in the time frame.

Even that would not be a "safe side" necessarily. If other Items rely on the UZSU Items (maybe by eval_trigger) those Items would get triggered at an unexpected time (the start time of SmartHomeNG).

Implementing those "missed" triggers would probably give more headache than it solves.

**Comment by onkelandy** [comment]
Imho the last entry of the uzsu would be enough, otherwise there might be quite some weird stuff going on. 

You are right with the triggers, so it might make sense to have an additional attribute like "uzsu_init: True" or something like that. If that is true, the last entry in the uzsu is set.

In my opinion it makes a lot of sense. 
UZSU activates a power socket with some ambient light. UZSU closes shutters in the evening. etc. etc. They'll stay open all night if smarthomeng wasn't running at the trigger time for whatever reason (maintenance, etc.)

**Comment by bmxp** [comment]
I am working on the orb/sunrise/sunset/ephem/uzsu/rrule/crontab complex to refactor things there and I will try to see if I can provide a solution to the above problem. But this won't happen to be fixed this year I think. Maybe for 1.5 by summertime with thorough testing.

**Comment by onkelandy** [comment]
Awesome! Count me in for testing ;)

**Comment by onkelandy** [comment]
@bmxp Did you already change anything in the "complex" you mentioned above? Actually I figured it is quite easy now to find previous values from UZSU in the current 1.5.1 version.
I'll push an update for the plugin soon, just wanted to know if you are still going to change anything there in the near future.

**Comment by bmxp** [comment]
No, I am very short of spare time and my thoughts to remove ephem in favor of skyfield or another library would need more time. So if you fix things or extend don't hesitate and go ahead.
